src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamespaceSymbol.cs

```diff
@@ -51,13 +51,23 @@ internal sealed override NamespaceExtent Extent
             }
         }
 
+        [PerformanceSensitive(
+            "https://github.com/dotnet/roslyn/issues/23582",
+            Constraint = "Provide " + nameof(ArrayBuilder<Symbol>) + " capacity to reduce number of allocations.",
+            AllowGenericEnumeration = false)]
```

Comment:
I think this is a bit overkill. Are we gonna add this attribute on all methods where we pass a parameter to GetInstance?

Reply:
> I think this is a bit overkill. Are we gonna add this attribute on all methods where we pass a parameter to GetInstance?

My current plan is to add this attribute whenever a change is indicated by a performance scenario which is not part of any of our automation. The URI (first parameter) is required, and points to a test case that revealed the original performance problem. For this specific use, see https://github.com/dotnet/roslyn/pull/23552#discussion_r155315714.

> PerformanceSensitive can't be use in code in PooledObjects namespace. The code in this namespace is imported by other repos via a source package.

The use in `ArrayBuilder<T>` is probably not essential for two reasons:

1. It doesn't use any of the enforceable properties of `PerformanceSensitiveAttribute`
2. Related issues with this type came up before and indicated that the scrutiny applied during reviews is substantial, so it's unlikely for a regression to make its way in

With that said, there is no build automation in place to prevent the use of `PerformanceSensitiveAttribute` in these types, so this may not be the last time this change occurs. Hopefully we can move the attribute to a code-aware library, allowing the projects that import pooled objects to also import the attribute.

Reply:
I think for this particular case it would be sufficient to have a comment by the GetInstance call ```// PERF: Critical to avoid extra allocations caused by expanding the array```.
We have many other places in the compiler where there are ```PERF``` related comments.
My 2 cents.